### ISSUE TITLE
Generator function stream is does not wait for inner asynchronous loop to finish

### ISSUE THREAD

**Issue by alsemenov** [dart-lang/sdk#24854]
Dart Programming Language Speciﬁcation version 1.9 reads:
>Let u be the stream associated with the immediately enclosing asynchronous
>for loop or generator function (9), if any. If another event e u of u occurs before
>execution of s is complete, handling of e u must wait until s is complete.

The following code tests this statement for generator function:
```
import 'dart:async';

Stream gen1(List log) async* {
  await for (int k in new Stream.fromIterable([1, 2])) {
    log.add('s'+k.toString());
    if (log.length<=4) {
      yield k.toString();
    }
    log.add('f'+k.toString());
  }
}

Future test1() async {
  List log = [];
  await for (String s in gen1(log)) {
    log.add(s);
  }
  return log;
}


main() {
  test1().then((log) => print("$log"));
}
```
The test produce the following output:
> [s1, 1, f1, s2, 2, f2]

The output shows that values '1' and '2' were handled in function test1() before the body of asynchronous for-in loop in function gen1() is finished.

Dart VM version: 1.12.2 (Wed Oct 21 07:10:52 2015) on "windows_x64"

**Comment by lrhn** [comment]
The behavior is correct and matches the specification.

What that part of the specification is saying is that if another event of `new Stream.fromIterable([1, 2])` occurs before the body of the `await for` loop is complete, it will be kept until the loop body completes. That is, event if the stream you are doing `await for` over spits out events as fast as possible, the body of the loop will only be for one at a time. This differs from a `stream.forEach((e) async { ... })` call which will call the callback as soon as possible.

It does not say that the events output by the async* function surrounding the loop are delayed. They can be handled immediately after the `yield` is executed because the async* function actually does yield control at the `yield` statement.

**Comment by alsemenov** [comment]
I am confused with the sentence 

>Let u be the stream associated with the immediately **enclosing** asynchronous
>for loop or generator function (9), **if any.**

Especially with the words 'enclosing' and 'if any'.
If you are correct, both of them are not right.
If there is single asynchronous for loop it should not be named as 'enclosed'
and remark 'if any' have no sense.

**Comment by lrhn** [comment]
The text is in the section about `await for` loops, but it is talking about a loop surrounding the current one (if any). So, if you have an await-for loop inside another await-for loop, the outer loop must not continue until the inner one is done. 

The general pattern in the spec is that any async operation inside a loop (`await`, `await for`, `yield`) states that a surrounding loop must pause until the body ends. If there is no async operation inside an await-for loop, then it's not required to pause between handling events.
You will find text similar to this in the section on `await`, and it means the same thing: that an enclosing async loop, if any, should not handle new events until the body is done.
(I can't find the same wording for `yield` or `yield*`, they just say that the function "may suspend").

**Comment by alsemenov** [comment]
Could you clarify the specification about the surrounding generator function instead of the loop?
I guess, that it should be the same:
if there is an await-for loop inside generator function, the stream related to generator function should not call event handlers till current iteration of the await-for loop finishes.

**Comment by lrhn** [comment]
There is no restriction on the events of a generator function. The generator function's stream subscription can deliver events immediately after a `yield`, whether the yield is in the body of the loop or not.

The only restriction is on loops internally in a single async function - the body of the loop is not reentrant, so any events happening on the await-for stream must be saved (or the stream paused) until the body is done.

**Comment by alsemenov** [comment]
I believe that the specification should be updated with clarifications:
* remove mentioning enclosing generator function
* add restriction, that body of the loop is not reentrant, and all events must be handled sequentially one by one.

Should I report new issue or it is possible to reuse this one?

**Comment by lrhn** [comment]
I don't know what Github protocol recommends, I have no problem with reusing this issue.

The restriction is already there, it just that it's implicit (async operations inside a loop states that the loop should be paused, instead of the loop stating that the body must be completed before the next event is processed).
